### PR TITLE
Refactors ambience to a subsystem

### PR DESCRIPTION
**Pull request by ghost**
:cl: Qustinnus
refactor: Ambience is now in a subsystem, and plays every now and then without you having to move to a new area for it to play
/:cl:

>stay in lavaland for 60 minutes
>never hear ambience because you are in the same area

stupid, I changed it. Now every area has a min and max cooldown and a subsystem automatically plays ambience when its time.

### CODE REVIEW

**Review by LemonInTheDark — APPROVED** [review]
Code looks good, not sure how I feel about the timers, we'll see how it works out I suppose

**Comment by Qustinnus** [comment]
They're not timers

**Comment by LemonInTheDark** [comment]
The delays between processes then?

**Comment by Qustinnus** [comment]
oh u meant design wise? It's basically the same as doing prob() except this guarantees a set time and lowers  the amount of proc calls to do.

File: code/controllers/subsystem/ambience.dm [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+	flags = SS_BACKGROUND|SS_NO_INIT
+	priority = FIRE_PRIORITY_AMBIENCE
+	runlevels = RUNLEVEL_GAME | RUNLEVEL_POSTGAME
+	wait = 1 SECONDS
+	///Assoc list of listening client - next ambience time
+	var/list/ambience_listening_clients = list()
+
+/datum/controller/subsystem/ambience/fire(resumed)
+	for(var/i in ambience_listening_clients)
+		var/client/client_iterator = i
```
Comment (Rohesie) [inline code/controllers/subsystem/ambience.dm:13]:
```suggestion
	for(var/client/client_iterator as anything in ambience_listening_clients)
```
New syntax.

File: code/modules/client/client_procs.dm [hunk 2]
```diff
@@ -634,6 +634,7 @@ GLOBAL_LIST_INIT(blacklisted_builds, list(
 	qdel(query_log_connection)
 
 	SSserver_maint.UpdateHubStatus()
+	SSambience.ambience_listening_clients -= src
```
Comment (Rohesie) [inline code/modules/client/client_procs.dm:637]: Are you sure this is the right place, instead of on `Destroy()` ?

Comment (Jared-Fogle) [reply code/modules/client/client_procs.dm:637]: it definitely is not, but what the hell is this proc

File: code/modules/client/client_procs.dm [hunk 3]
```diff
@@ -1099,3 +1101,9 @@ GLOBAL_LIST_INIT(blacklisted_builds, list(
 		winset(src, "mapwindow.map", "right-click=false")
 		winset(src, "default.Shift", "is-disabled=true")
 		winset(src, "default.ShiftUp", "is-disabled=true")
+
+/client/proc/update_ambience_pref()
+	if(prefs.toggles & SOUND_AMBIENCE)
+		SSambience.ambience_listening_clients[src] = world.time + 10 SECONDS //Just wait 10 seconds before the next one aight mate? cheers.
```
Comment (Rohesie) [inline code/modules/client/client_procs.dm:1107]:
```suggestion
	if(prefs.toggles & SOUND_AMBIENCE)
		if(SSambience.ambience_listening_clients[src] > world.time)
			return // If already properly set we don't want to reset the timer.
		SSambience.ambience_listening_clients[src] = world.time + 10 SECONDS //Just wait 10 seconds before the next one aight mate? cheers.
```
Makes sense?

**Pull request merged**